src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs

```diff
@@ -15,44 +16,72 @@ namespace Microsoft.Extensions.Configuration.AzureKeyVault
     /// </summary>
     internal class AzureKeyVaultConfigurationProvider : ConfigurationProvider
     {
+        private readonly bool _reloadOnChange;
+        private readonly int _reloadPollDelay;
         private readonly IKeyVaultClient _client;
         private readonly string _vault;
         private readonly IKeyVaultSecretManager _manager;
+        private Dictionary<string, SecretAttributes> _loadedSecrets;
+        private Task _pollingTask;
+        private CancellationTokenSource _cancellationToken;
 
         /// <summary>
         /// Creates a new instance of <see cref="AzureKeyVaultConfigurationProvider"/>.
         /// </summary>
         /// <param name="client">The <see cref="KeyVaultClient"/> to use for retrieving values.</param>
         /// <param name="vault">Azure KeyVault uri.</param>
         /// <param name="manager"></param>
-        public AzureKeyVaultConfigurationProvider(IKeyVaultClient client, string vault, IKeyVaultSecretManager manager)
+        /// <param name="reloadOnChange">Whether the configuration should be reloaded if the Azure KeyVault changes.</param>
+        /// <param name="reloadPollDelay">Number of milliseconds to wait inbetween each attempt at polling the Azure KeyVault for changes.</param>
+        public AzureKeyVaultConfigurationProvider(IKeyVaultClient client, string vault, IKeyVaultSecretManager manager, bool reloadOnChange = false, int reloadPollDelay = 10000)
         {
-            if (client == null)
-            {
-                throw new ArgumentNullException(nameof(client));
-            }
-            if (vault == null)
-            {
-                throw new ArgumentNullException(nameof(vault));
-            }
-            if (manager == null)
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _vault = vault ?? throw new ArgumentNullException(nameof(vault));
+            _manager = manager ?? throw new ArgumentNullException(nameof(manager));
+
+            if (reloadOnChange && reloadPollDelay <= 0)
             {
-                throw new ArgumentNullException(nameof(manager));
+                throw new ArgumentException("{0} must be greater than 0", nameof(reloadPollDelay));
             }
 
-            _client = client;
-            _vault = vault;
-            _manager = manager;
+            _pollingTask = null;
+            _cancellationToken = new CancellationTokenSource();
+            _reloadOnChange = reloadOnChange;
+            _reloadPollDelay = reloadPollDelay;
+            _loadedSecrets = new Dictionary<string, SecretAttributes>();
+        }
+
+        ~AzureKeyVaultConfigurationProvider()
+        {
+            _cancellationToken.Cancel();
         }
 
         public override void Load() => LoadAsync().ConfigureAwait(false).GetAwaiter().GetResult();
 
+        private async Task PollForSecretChangesAsync(int reloadPollDelay, CancellationToken ct)
```

Comment:
This method would reload all keys twice. Once to check if reload is required another to actually reload them. I think reload logic can be merged into LoadAsync method that would call              OnReload(); only if new keys/versions of keys are detected.